src/main/scala/com/fulcrumgenomics/commons/collection/ParIterator.scala

```diff
@@ -0,0 +1,147 @@
+
+import com.fulcrumgenomics.commons.CommonsDef._
+import com.fulcrumgenomics.commons.async.AsyncIterator
+
+import java.util.concurrent.ForkJoinPool
+import scala.collection.AbstractIterator
+import scala.collection.parallel.{ForkJoinTaskSupport, TaskSupport}
+
+/**
+  * Companion object to help with manufacturing ParIterators.
+  */
+object ParIterator {
+  /**
+    * Constructs a ParIterator from the underlying iterator, that will perform operations in parallel on chunks
+    * of `chunkSize` using `threads` threads.
+    *
+    * @param iter the underlying iterator to parallelize over
+    * @param chunkSize the number of elements to chunk together for parallel processing
+    * @param threads the number of threads to use for parallel computations like `map()`
+    * @param chunkBuffer if Some(n) use an [[AsyncIterator]] to pre-buffer n chunks from the input
+    *                    read for parallel computation
+    */
+  def apply[A](iter: Iterator[A], chunkSize: Int = 1024, threads: Int, chunkBuffer: Option[Int] = Some(2)): ParIterator[A] = {
+    val pool = new ForkJoinPool(threads, ForkJoinPool.defaultForkJoinWorkerThreadFactory, null, true)
+    val support = new ForkJoinTaskSupport(pool)
+    apply(iter, chunkSize, support, chunkBuffer)
+  }
+
+  /**
+    * Constructs a ParIterator from the underlying iterator, that will perform operations in parallel on chunks
+    * of `chunkSize` using the provided `TaskSupport`.
+    *
+    * @param iter the underlying iterator to parallelize over
+    * @param chunkSize the number of elements to chunk together for parallel processing
+    * @param support the TaskSupport to use for parallel computations like `map()`
+    * @param chunkBuffer if Some(n) use an [[AsyncIterator]] to pre-buffer n chunks from the input
+    *                    read for parallel computation
+    */
+  def apply[A](iter: Iterator[A], chunkSize: Int, support: TaskSupport, chunkBuffer: Option[Int]): ParIterator[A] = {
+    val grouped = {
+      val g = iter.grouped(chunkSize)
+      chunkBuffer match {
+        case None    => g
+        case Some(n) => AsyncIterator(g, Some(n))
+      }
+    }
+
+    new ParIterator[A](grouped, chunkSize, support)
+  }
+}
+
+
+/**
+  * A parallel iterator that operates by grouping or chunking the underlying iterator and then performing
+  * operations on the chunks in parallel.  Optionally supports wrapping the iterator in an [[AsyncIterator]]
+  * to draw items through the parallel processing and make them available for consumption while the next chunk(s)
+  * are processed.
+  *
+  * For example:
+  *
+  * {{{
+  * val iter = Io.readLines(reallyBigFile)
+  * ParIterator(iter, chunkSize=4096, threads=8)
+  *   .filter(_.contains("foo"))
+  *   .map(_.toUpperCase)
+  *   .map(_.reverse)
+  *   .toAsync(cache=4096 * 8)
+  *   .foreach(println)
+  * }}}
+  */
+class ParIterator[A] private (private val iter: Iterator[Seq[A]],
+                              val chunkSize: Int = 1024,
```

Comment:
defaults could be moved into the companion object.  Looks like this default is used in a few places.

Reply:
Done